_posts/2024-08-06-how-to-talk-about-CHERI.markdown

```diff
@@ -0,0 +1,185 @@
+---
+layout: post
+title:  "How to talk to your parents about hardware memory safety"
+date:   2024-08-06
+categories: cheri
+author: David Chisnall
+---
+
+Some conversations are difficult to have with members of older generations who grew up with different social norms.
+In particular, when you're talking to people who grew up with PDP-11s with their completely flat memory, or Lisp machines or Burroughs Large Systems with their deeply opinionated and language-integrated hardware memory safety, you may find it hard to talk about CHERI.
+This guide aims to help you have those conversations with the minimum of stress on both sides.
+
+## You don't understand me or my object model!
+
+When we talk informally about CHERI, it's tempting to say things like 'CHERI provides memory safety' or 'CHERI gives you control-flow integrity'.
+The CHERI project started 14 years ago and people who have been working on it for a decade or so know that this is a shorthand but when you're engaging with people who haven't yet accepted the truth of capability systems into their life, it's important to be precise.
+
+A CPU instruction set architecture (ISA), or an ISA extension, for general-purpose computing, cannot provide memory safety.
+A definition of memory safety starts from a definition of an object model.
+C has an object model, as has Rust, Java, and so on.
+Each object model defines the bounds of objects and their lifetime.
+If you access an object after the end of its lifetime, or after you try to access outside of the bounds of an object, you have violated memory safety.
+A CPU can't provide memory safety because it doesn't know what the object model for the running program is.
+
+The same is true for control-flow integrity (CFI).
+CFI defines a program as a directed graph of of blocks of instructions with arcs in between them representing things like function calls, returns, and so on.
+Again, this graph is a property of the language.
+For example, in C++ if you call a virtual function then the object on which it's called must be a subclass of the class on which the function is defined.
+This set of properties can be quite rich and so most CFI schemes focus on preventing particularly dangerous invalid transitions, rather than preventing *all* invalid transitions.
+For example, a C CFI scheme may allow you to call `system` when you meant `puts`, because they take the same argument type, but prevent you from calling `fputs`.
+
+## You're in my language and you'll follow my language's abstract machine's rule!
+
+So what's the point of CHERI if an ISA can't give you these properties?
+CHERI is not designed to give you an object model that you must conform to, it's designed to *give language implementers tools to enforce these properties*.
```

Comment:
It does give language implements a lot to work with - but I think a lot of the value also comes from that fact that those same features are exposed to programmers to be able to explicitly create read-only pointers when they need to, etc

Otherwise it sound a bit as if you'r only writing this blog for  language implementers